### PR TITLE
Sprite Sheet Support, Scene Viewport, Test Editor Project.

### PR DESCRIPTION
**Pull request by TejasJD**
# Sprite Sheet Support
#66 Is the initial PR for sprite sheet support. But since, I did changes that might invalidate the changes related to that PR and added new changes on top of that, I decided to close it, change the branch name, commit new changes and open up a new with those changes as well.
# Scene Viewport
- Dockable viewport.
image
- Updates the aspect ratio of the camera, upon changing the size of the viewport.
image

### CODE REVIEW

File: HorizonEditor/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+set(HZNEDITOR_SOURCE_FILES
+	"src/Main.cpp"
+  "src/EditorLayer.cpp")
+
+set(HZNEDITOR_HEADER_FILES
+ "include/EditorLayer.h" )
+
+# 'HznApplication' target which would generate an executable.
+add_executable(HznEditor ${HZNEDITOR_SOURCE_FILES} ${HZNEDITOR_HEADER_FILES})
```
Comment (LiamANeeson) [inline HorizonEditor/CMakeLists.txt:14]: Executable for the Game/Scene?

File: HorizonEditor/src/EditorLayer.cpp [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+	m_AspectRatio((float)Hzn::App::getApp().getAppWindow().getWidth() / (float)Hzn::App::getApp().getAppWindow().getHeight()),
+	m_CameraController(Hzn::OrthographicCameraController(m_AspectRatio, 1.0f))
+{
+}
+
+void EditorLayer::onAttach()
+{
+	HZN_TRACE("Editor Layer Attached!");
+	m_CheckerboardTexture = Hzn::Texture2D::create("assets/textures/bear.png");
+	Hzn::FrameBufferProps props;
```
Comment (LiamANeeson) [inline HorizonEditor/src/EditorLayer.cpp:15]: What are props in the context of FrameBuffers

File: HorizonEditor/src/EditorLayer.cpp [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+
+void EditorLayer::onUpdate(Hzn::TimeStep ts)
+{
+	m_CameraController.onUpdate(ts);
+	m_FrameBuffer->bind();
+	Hzn::RenderCall::setClearColor({ 0.1f, 0.1f, 0.1f, 1.0f });
+	Hzn::RenderCall::submitClear();
+	Hzn::Renderer2D::beginScene((const Hzn::OrthographicCamera&)m_CameraController.getCamera());
+
+	for (int32_t i = 0; i < quads; ++i)
```
Comment (LiamANeeson) [inline HorizonEditor/src/EditorLayer.cpp:32]: What are the quads?

File: HorizonEditor/src/EditorLayer.cpp [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+    static bool opt_fullscreen = true;
+    static bool opt_padding = false;
+    static ImGuiDockNodeFlags dockspace_flags = ImGuiDockNodeFlags_None;
+
+    // We are using the ImGuiWindowFlags_NoDocking flag to make the parent window not dockable into,
+    // because it would be confusing to have two docking targets within each others.
+    ImGuiWindowFlags window_flags = ImGuiWindowFlags_MenuBar | ImGuiWindowFlags_NoDocking;
+    if (opt_fullscreen)
+    {
+        const ImGuiViewport* viewport = ImGui::GetMainViewport();
```
Comment (LiamANeeson) [inline HorizonEditor/src/EditorLayer.cpp:82]: This looks good

File: HorizonEditor/CMakeLists.txt (hunk 1 above)
Comment (TejasJD) [reply HorizonEditor/CMakeLists.txt:14]: This is the standalone editor that I am writing to test code.. and develop things.

File: HorizonEditor/src/EditorLayer.cpp (hunk 3 above)
Comment (TejasJD) [reply HorizonEditor/src/EditorLayer.cpp:32]: These quads, it's just a variable that indicates the side of the test grid that I am rendering.

File: HorizonEditor/src/EditorLayer.cpp (hunk 4 above)
Comment (TejasJD) [reply HorizonEditor/src/EditorLayer.cpp:82]: This is the dockspace code from ImGui docking branch.

File: HorizonEditor/src/EditorLayer.cpp [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+    // When using ImGuiDockNodeFlags_PassthruCentralNode, DockSpace() will render our background
+    // and handle the pass-thru hole, so we ask Begin() to not render a background.
+    if (dockspace_flags & ImGuiDockNodeFlags_PassthruCentralNode)
+        window_flags |= ImGuiWindowFlags_NoBackground;
+
+    // Important: note that we proceed even if Begin() returns false (aka window is collapsed).
+    // This is because we want to keep our DockSpace() active. If a DockSpace() is inactive,
+    // all active windows docked into it will lose their parent and become undocked.
+    // We cannot preserve the docking relationship between an active window and an inactive docking, otherwise
+    // any change of dockspace/settings would lead to windows being stuck in limbo and never being visible.
```
Comment (LiamANeeson) [inline HorizonEditor/src/EditorLayer.cpp:105]: Good work

File: HorizonEngine/Platform/OpenGL/GLFrameBuffer.cpp [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+		if(m_FrameBufferId != 0)
+		{
+			destroy();
+		}
+
+		glGenFramebuffers(1, &m_FrameBufferId);
+		glBindFramebuffer(GL_FRAMEBUFFER, m_FrameBufferId);
+
+		// Color attachment
+		glGenTextures(1, &m_ColorAttachment);
```
Comment (LiamANeeson) [inline HorizonEngine/Platform/OpenGL/GLFrameBuffer.cpp:39]: If the GLFrameBuffer is being invalidated why are we attaching colours, textures etc. with in this invalidation

Comment (TejasJD) [reply HorizonEngine/Platform/OpenGL/GLFrameBuffer.cpp:39]: Because we want to resubmit that data, when we are recreating the framebuffer. Frame buffers need to have color and depth attachments for rendering.

File: HorizonEditor/src/EditorLayer.cpp (hunk 3 above)
Comment (alex-mulkerrins) [reply HorizonEditor/src/EditorLayer.cpp:32]: The 'quads' reaches scope from EditorLayer.h right?

File: HorizonEngine/Platform/OpenGL/GLWindow.cpp [hunk 7]
```diff
@@ -52,77 +52,78 @@ namespace Hzn
 		//! sets callback to the window close event
 		glfwSetWindowCloseCallback(m_Window, [](GLFWwindow* window)
 			{
-				WindowData data = *(WindowData*)glfwGetWindowUserPointer(window);
+				WindowData* data = (WindowData*)glfwGetWindowUserPointer(window);
 				WindowCloseEvent eventObj;
-				data.callback(eventObj);
+				data->callback(eventObj);
```
Comment (LiamANeeson) [inline HorizonEngine/Platform/OpenGL/GLWindow.cpp:57]: Why did you change the data callbacks from accessing via an instance( . ) to accessing them via a pointer (->)

File: HorizonEditor/src/EditorLayer.cpp (hunk 3 above)
Comment (TejasJD) [reply HorizonEditor/src/EditorLayer.cpp:32]: Quads are scope between 5 to 1000, if you see the ImGui slider code, that is what changes this variable.

**Review by Minghui-Gao — APPROVED** [review]

**Review by LiamANeeson — COMMENTED** [review]
All this work looks great, have some general questions but really good work. I see you have started using the summary tags for the documentation also this is very good and I also found those comments and summary sections very useful when going through the PR.

File: HorizonEngine/src/HorizonEngine/Renderer/Sprite.h [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+		/// is (width of the sheet, height of the sheet). </param>
+		/// <param name="cellSize">Dimensions of the grid cells of the sprite sheet.</param>
+		/// <param name="spriteSize">Dimensions of the sprite.</param>
+		/// <returns>Shared pointer to the 2D Sprite.</returns>
+		static std::shared_ptr<Sprite2D> create(
+			const std::shared_ptr<Texture2D>& spriteSheet,
+			const glm::vec2& sheetPosition,
+			const glm::vec2& cellSize,
+			const glm::vec2& spriteSize = { 1.0f, 1.0f }
+		);
```
Comment (alex-mulkerrins) [inline HorizonEngine/src/HorizonEngine/Renderer/Sprite.h:34]: Why is shared_ptr used here? Is it so that ownership of the sprite created is shared and can be accessed by different parts?

File: HorizonEngine/src/HorizonEngine/Renderer/Sprite.h [hunk 9]
```diff
@@ -0,0 +1,60 @@
+#pragma once
+
+#ifndef HZN_SPRITE_H
+#define HZN_SPRITE_H
+
+#include "Texture.h"
+
+namespace Hzn
+{
+	/// <summary>
```
Comment (alex-mulkerrins) [inline HorizonEngine/src/HorizonEngine/Renderer/Sprite.h:10]: I like the use of documentation throughout this PR, makes it easier to understand, cheers!

**Review by alex-mulkerrins — APPROVED** [review]

**Review by prajakta-kulkarni-git — APPROVED** [review]

File: HorizonEngine/src/HorizonEngine/Renderer/Sprite.h (hunk 8 above)
Comment (TejasJD) [reply HorizonEngine/src/HorizonEngine/Renderer/Sprite.h:34]: Hi,
its just that if you want to pass this sprite to a function and such, you can easily do so, without dealing with move semantics of passing unique_ptr through functions.
